hub3/ead/dao.go

```diff
@@ -0,0 +1,435 @@
+package ead
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/delving/hub3/config"
+	"github.com/delving/hub3/hub3/ead/eadpb"
+	"github.com/delving/hub3/hub3/fragments"
+	"github.com/delving/hub3/ikuzo/domain/domainpb"
+	"github.com/delving/hub3/ikuzo/service/x/index"
+	rdf "github.com/kiivihal/rdf2go"
+	"github.com/rs/xid"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	daoPath = "dao"
+)
+
+type DaoClient struct {
+	bi           *index.Service
+	client       *http.Client
+	HttpFallback bool // retrieve DAO url if not present locally
+}
+
+func NewDaoClient(bi *index.Service) DaoClient {
+	return DaoClient{
+		client: &http.Client{Timeout: 10 * time.Second},
+		bi:     bi,
+	}
+}
+
+func (c *DaoClient) dropOrphans(cfg DaoConfig) error {
+	m := &domainpb.IndexMessage{
+		OrganisationID: cfg.OrgID,
+		DatasetID:      cfg.ArchiveID,
+		Revision: &domainpb.Revision{
+			SHA:     cfg.UUID,
+			Path:    cfg.RevisionKey,
+			GroupID: cfg.InventoryID,
+		},
+		ActionType: domainpb.ActionType_DROP_ORPHANS,
+	}
+
+	// publish message
+	if err := c.bi.Publish(context.Background(), m); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *DaoClient) PublishFindingAid(cfg DaoConfig) error {
+	fa, err := cfg.FindingAid(c)
+	if err != nil {
+		return err
+	}
+
+	cfg.updateRevisionKey()
+
+	for _, file := range fa.Files {
+		fg, err := cfg.fragmentGraph(file)
+		if err != nil {
+			return err
+		}
+		fg.Meta.SourceID = cfg.RevisionKey
+
+		m, err := fg.IndexMessage()
+		if err != nil {
+			return err
+		}
+
+		if c.bi != nil {
+			c.bi.Publish(context.Background(), m)
+		}
+	}
+
+	fg, err := cfg.findingAidFragmentGraph(&fa)
+	if err != nil {
+		return err
+	}
+	fg.Meta.SourceID = cfg.RevisionKey
+
+	m, err := fg.IndexMessage()
+	if err != nil {
+		return err
+	}
+
+	if c.bi != nil {
```

Comment:
Moet de c.bi service niet altijd aanwezig zijn? Zo ja dan zou ik de check bovenaan de functie plaatsen en een error returnen als het er niet is.